### DISCUSSION TITLE
Can CsvHelper change to netstandard2.1 only?

### DISCUSSION THREAD

**Discussion by JoshClose** [JoshClose/CsvHelper#1474] [General]
The next version of the parser uses `Span<T>` and `Memory<T>` and it more than 5 times faster. These are only available in `netstandard2.1` though. I'd like a discussion on how this should be handled.

Here are the options as I see them. Feel free to suggest another option if you have one.

1. Drop everything but `netstandard2.1`.
  This is the option I'd like to implement. You'll need to convince me why I shouldn't do this. 
    * Users that aren't on `netstandard2.1` can still use the old packages. They just won't get new features. 
    * Bugs fixes could still be implemented into older releases.
    * Only one version to support.
    * No need for `#if` preprocessor directives.
    * There could possibly be issues with NuGet trying to add a `netstandard2.1` only package to projects that aren't `netstandard2.1`.
1. Create another NuGet package.
    * I'd really like `CsvHelper` to always be the latest version, so this is not appealing.
    * I don't want to maintain 2 packages.
1. Offer as an opt-in secondary parser.
    * There would be `CsvStackParser` or something like that. `CsvParser` would then be deprecated and removed later. When the removal happens, I would want `CsvStackParser` to be renamed to `CsvParser`. This could all get confusing.
    * `CsvParser` could be the new `nestandard2.1` only parser and the current parser could be renamed to `CsvLegacyParser`. I like this route better, but will break most people. Breaking doesn't really concern me all that much though because it will be a major version change.
    * Unit tests would somehow need to be updated to run both parsers.
1. Use the new parser for `netstandard2.1` and the old for everything else.
    * This would hide bugs and confuse people.

**Comment by StefanBertels** [comment]
I vote for (1), i.e. netstandard2.1 only.

**Comment by PatrickHofman** [comment]
Option (1) definitely. Take the pain once, also for the users implementing this and wanting new features. Everyone else, stay on the old version.

The performance gain is impressive. Good work!

**Comment by monty241** [comment]
.NET standard is a set of api's as far as I know, implemented in some piece of software. Since all .NET Core releases have already reached end of support except 2.1 and 3.1, I would recommend sticking to 2.1.

So option (1)

See https://dotnet.microsoft.com/platform/support/policy/dotnet-core

**Comment by jongleur1983** [comment]
We use a quite old version (and are fine with that) as nobody cared to update it while checking if nothing breaks.
As we didn't update for a long time, I'd support option (1) as well - once we can switch to Core (in the WPF world) we'll have to tackle the upgrade nevertheless, and it's unlikely we'll do it earlier.

**Comment by mudnug** [comment]
Does anyone know if there is a tool that can list all opensource projects that depend on CsvHelper  and show what percent support netstandard2.1?

**Comment by PatrickHofman** [comment]
@mudnug I don't see the relevance of that, since those projects can still use the old version if they want to. There is no need to update. And all packages eventually upgrade their framework, so no need to be afraid of breaking anyone's project.

**Comment by JoshClose** [comment]
I agree with @PatrickHofman but I found these 2 sites.

https://nugetmusthaves.com/Dependency/CsvHelper

https://libraries.io/nuget/CsvHelper/dependents

**Comment by c0shea** [comment]
I'm ok with option 1 so long as bugs that apply to the latest non-.netstandard 2.1 version are backported. We're on .NET Framework 4.8.

**Comment by JoshClose** [comment]
If they apply to both, then they would be.

**Comment by istrupin** [comment]
Option 1 with backported bug fixes sounds ideal and would definitely  get my vote.

**Comment by twelsh920** [comment]
Any update on this? Really looking forward to the potential performance improvements you've mentioned elsewhere that use the newer `Span<T>`, `Memory<T>` apis

**Reply by JoshClose** [reply]
This was done in version 19. The speed improvement was huge and it's not using span/memory. Still compatible with .NET 4.5

https://www.joelverhagen.com/blog/2020/12/fastest-net-csv-parsers

**Comment by JoshClose** [comment]
It's going to be a bit still. This feature requires refactoring of most things.
